sentinel-dashboard/src/main/webapp/resources/app/scripts/controllers/gateway/flow.js

```diff
@@ -0,0 +1,232 @@
+          } else {
+            $scope.rules = [];
+            $scope.rulesPageConfig.totalCount = 0;
+          }
+        });
+    };
+    $scope.getMachineRules = getMachineRules;
+
+
+    getApiNames();
+    function getApiNames() {
+      if (!$scope.macInputModel) {
+        return;
+      }
+
+      var mac = $scope.macInputModel.split(':');
+      GatewayApiService.queryApis($scope.app, mac[0], mac[1]).success(
+        function (data) {
+          if (data.code == 0 && data.data) {
+            $scope.apiNames = [];
+
+            data.data.forEach(function (api) {
+              $scope.apiNames.push(api["apiName"]);
+            });
+          }
+        });
+    }
+
+    $scope.intervalUnits = [{val: 0, desc: '秒'}, {val: 1, desc: '分'}, {val: 2, desc: '时'}, {val: 3, desc: '天'}];
+
+    var gatewayFlowRuleDialog;
+    $scope.editRule = function (rule) {
+      $scope.currentRule = angular.copy(rule);
+      $scope.gatewayFlowRuleDialog = {
+        title: '编辑流控规则',
+        type: 'edit',
+        confirmBtnText: '保存'
+      };
+      gatewayFlowRuleDialog = ngDialog.open({
+        template: '/app/views/dialog/gateway/flow-rule-dialog.html',
+        width: 780,
+        overlay: true,
+        scope: $scope
+      });
+    };
+
+    $scope.addNewRule = function () {
+      var mac = $scope.macInputModel.split(':');
+      $scope.currentRule = {
+        grade: 0,
+        app: $scope.app,
+        ip: mac[0],
+        port: mac[1],
+        resourceMode: 0,
+        interval: 1,
+        intervalUnit: 0,
+        controlBehavior: 0,
+        burst: 0,
+        maxQueueingTimeoutMs: 0
+      };
+      $scope.gatewayFlowRuleDialog = {
+        title: '新增流控规则',
+        type: 'add',
+        confirmBtnText: '新增'
+      };
+      gatewayFlowRuleDialog = ngDialog.open({
+        template: '/app/views/dialog/gateway/flow-rule-dialog.html',
+        width: 780,
+        overlay: true,
+        scope: $scope
+      });
+    };
+
+    $scope.saveRule = function () {
+      if (!GatewayFlowService.checkRuleValid($scope.currentRule)) {
+        return;
+      }
+      if ($scope.gatewayFlowRuleDialog.type === 'add') {
+        addNewRule($scope.currentRule);
+      } else if ($scope.gatewayFlowRuleDialog.type === 'edit') {
+        saveRule($scope.currentRule, true);
+      }
+    };
+
+    $scope.useParamItem = function () {
+      $scope.currentRule.paramItem = {
+        parseStrategy: 0,
+        matchStrategy: 0
+      };
+    };
+
+    $scope.notUseParamItem = function () {
+      $scope.currentRule.paramItem = null;
+    };
+
+    function addNewRule(rule) {
+      GatewayFlowService.newRule(rule).success(function (data) {
+        if (data.code == 0) {
+          getMachineRules();
+          gatewayFlowRuleDialog.close();
+        } else {
+          alert('新增流控规则失败!' + data.msg);
+        }
+      });
+    };
+
+    function saveRule(rule, edit) {
+      GatewayFlowService.saveRule(rule).success(function (data) {
```

Comment:
Maybe we need to check whether the rule is valid here?

Reply:
In 115 line, before calling `addNewRule` and `saveRule` , call `GatewayFlowService.checkRuleValid($scope.currentRule)` to check, and the check method has been improved.